github/downloader_test.go

```diff
@@ -80,74 +118,380 @@ func getDownloader() (*Downloader, *testutils.Memory, error) {
 		return nil, nil, err
 	}
-
 	storer := new(testutils.Memory)
 	downloader.storer = storer
 	return downloader, storer, nil
 }
 
-func testOnlineRepo(t *testing.T, oracle RepositoryTest, d *Downloader, storer *testutils.Memory) {
+// Tests:
+// 1. Online (live) download with token for git-fixtures org and memory cache
+// 2. Online (live) download with token for git-fixtures repos and memory cache
+// 4. Offline (recorded) download without token for src-d/gitbase repo and memory cache
+// 3. Offline (recorded) download without token for src-d org and memory cache
+// 5. Online (live) download with token for git-fixtures org and DB
+// 6. Online (live) download with token for git-fixtures repos and DB
+// 7. Offline (recorded) download without token for src-d org and DB
+// 8. Offline (recorded) download without token for src-d/gitbase repo and DB
+
+type DownloaderTestSuite struct {
+	suite.Suite
+	db         *sql.DB
+	downloader *Downloader
+}
+
+func (suite *DownloaderTestSuite) SetupSuite() {
+	if !isOSXOnTravis() {
+		suite.db = getDB(suite.T())
+		suite.NotNil(suite.db)
+	}
+}
+
+// TestOnlineRepositoryDownload Tests the download of known and fixed GitHub repositories
+func (suite *DownloaderTestSuite) TestOnlineRepositoryDownload() {
+	t := suite.T()
+	checkToken(t)
+	tests, err := loadTests("../testdata/online-repository-tests.json")
+	suite.NoError(err, "Failed to read the testcases")
+	downloader, storer, err := getDownloader()
+	suite.NoError(err, "Failed to instantiate downloader")
+	for _, test := range tests.RepositoryTests {
+		test := test // pinned, see scopelint for more info
+		t.Run(fmt.Sprintf("Repo: %s/%s", test.Owner, test.Repository), func(t *testing.T) {
+			testRepo(t, test, downloader, storer, false)
+		})
+	}
+}
+
+func testRepo(t *testing.T, oracle testutils.RepositoryTest, d *Downloader, storer *testutils.Memory, strict bool) {
 	err := d.DownloadRepository(context.TODO(), oracle.Owner, oracle.Repository, oracle.Version)
 	require := require.New(t) // Make a new require object for the specified test, so no need to pass it around
 	require.Nil(err)
-	// Sample some properties that will not change, no topics available in git-fixtures
 	require.Equal(oracle.URL, storer.Repository.Url)
 	require.Equal(oracle.CreatedAt, storer.Repository.CreatedAt.String())
 	require.Equal(oracle.IsPrivate, storer.Repository.IsPrivate)
 	require.Equal(oracle.IsArchived, storer.Repository.IsArchived)
 	require.Equal(oracle.HasWiki, storer.Repository.HasWikiEnabled)
-	require.Len(storer.PRs, oracle.NumOfPRs)
-	require.Len(storer.PRComments, oracle.NumOfPRComments)
+	require.ElementsMatch(oracle.Topics, storer.Topics)
+	numOfPRReviews, numOfPRReviewComments := storer.CountPRReviewsAndReviewComments()
+	if strict {
+		require.Len(storer.PRs, oracle.NumOfPRs)
+		require.Len(storer.PRComments, oracle.NumOfPRComments)
+		require.Len(storer.Issues, oracle.NumOfIssues)
+		require.Len(storer.IssueComments, oracle.NumOfIssueComments)
```

Comment:
nitpick: could you invert the arguments just for consistency/readability?

Reply:
The function signature for `Len` is: `func Len(t TestingT, object interface{}, length int, msgAndArgs ...interface{})` so the slice has to come in front
on the other hand, the signature for `Equals` is:
`func Equal(t TestingT, expected interface{}, actual interface{}, msgAndArgs ...interface{})`
so the expected interface is in front and the actual is second.

I didn't like it as well, but I thought it is better to go with the library docs.

Reply:
oh ok. 

> I didn't like it as well, but I thought it is better to go with the library docs.

sure!